base/channels.jl

```diff
@@ -183,7 +183,7 @@ Close a channel. An exception (optionally given by `excp`), is thrown by:
 * [`put!`](@ref) on a closed channel.
 * [`take!`](@ref) and [`fetch`](@ref) on an empty, closed channel.
 """
-function close(c::Channel, excp::Exception=closed_exception())
+function close(c::Channel, @nospecialize(excp::Exception=closed_exception()))
```

Comment:
> ~Avoids dynamic dispatch when closing a Channel with an Exception~

Oh, jk i forgot that `Channel` is also a parametric type, so this will still have a dispatch to do.

... is there any reason to specialize on the Channel type? None of the Channel fields accessed in this function depend on the type of the parameter.. Would julia be able to tell that, and still be able to produce concrete code for a type without its type parameter if none of the accessed fields depend on the parameter? My guess is the answer is **no**, but it would be a neat feature.

@vchuravy or @pchintalapudi you are probably the best people I could get to answer that question - do you know?

Reply:
@quinnj and @vtjnash: Without the above, i guess this will still trigger compilation, due to the dispatch on the Channel type.

Can I mark that as nospecialize too?

Reply:
You may incur internal dispatch (for lock and unlock), but it can potentially still figure out memory layout and dispatch specializations

Reply:
It looks like inference should know the type of Channel. Only the exception argument is type Any

Reply:
> You may incur internal dispatch (for lock and unlock), but it can potentially still figure out memory layout and dispatch specializations

Oh right. Good points..

.... Is there any reason I can't mark these as `@nospecialize(c::Channel)` as well???:
https://github.com/JuliaLang/julia/blob/36ace5430d5b54ec3f7045b980002fe2f600e508/base/channels.jl#L525-L528

Reply:
...assuming that julia can still produce good code since `cond_take` doesn't depend on `{T}`?

Reply:
...I'm finding that I can't rely on Cthulhu to answer this question, since it seems to disregard the `@nospecialize`, and shows me a function that's been inferred specifically for the type i'm calling with, even though codegen will ultimately produce a single generic native function, right?

> It looks like inference should know the type of Channel. Only the exception argument is type Any

@vtjnash: I'm not sure what you meant by this message. Can you expand on it? How does it apply to the situation where I mark both arguments nospecialize in close?

Reply:
Ohhhhhh you're saying I don't _need_ to mark it nospecialize in order to avoid a dispatch, since inference already knows the channel type, so i could still avoid the dispatch even if the exception type is unknown! 💡 great point. So it could still be an invoke, or even inlined 👍 

Neat!

Oh, so then maybe Cthulhu was right after all, because it was devirtualizing my call!  Aha, yes that was it. Neat!! :)
Julia is pretty cool you guys.

So then: the only fear would be that if this is the **first exception** we've ever called `close()` on for this channel type, we might still stackoverflow. But if we've done it for some other exceptions first, we'll be okay.
... that's probably still an improvement, and so we can probably do without marking it `@nospecialize`. (Unless we think it would still produce good code even in that case.)

Reply:
Wow, incredible! Maybe it actually _can_ produce good code, even if the Channel is nospecialize!
```julia
julia> struct Foo{T}
           a::Int
           b::T
       end

julia> function foo(@nospecialize(x::Foo))
           bar(x.a)
       end
foo (generic function with 1 method)

julia> bar(v) = v + 1
bar (generic function with 1 method)

julia> function run(v::Vector{Any})
           return foo(v[1])
       end
run (generic function with 1 method)

julia> using Cthulhu

julia> @descend run(Any[Foo(1, "hi")])
run(v::Vector{Any}) in Main at REPL[4]:1
1 function run(v::Vector{Any}::Vector{Any})::Int64
2     return foo(v::Vector{Any}[1]::Any)::Int64
3 end
Select a call to descend into or ↩ to ascend. [q]uit. [b]ookmark.
Toggles: [w]arn, [h]ide type-stable statements, [t]ype annotations, [s]yntax highlight for Source/LLVM/Native.
Show: [S]ource code, [A]ST, [T]yped code, [L]LVM IR, [N]ative code
Actions: [E]dit source code, [R]evise and redisplay
   2 v::Vector{Any}[1]
 • 2 foo(v::Vector{Any}[1]::Any)
   ↩
foo(x::Foo) in Main at REPL[2]:1
1 function foo(@nospecialize(x::Foo::Foo))::Int64
2     bar(x::Foo.a::Int64)::Int64
3 end
Select a call to descend into or ↩ to ascend. [q]uit. [b]ookmark.
Toggles: [w]arn, [h]ide type-stable statements, [t]ype annotations, [s]yntax highlight for Source/LLVM/Native.
Show: [S]ource code, [A]ST, [T]yped code, [L]LVM IR, [N]ative code
Actions: [E]dit source code, [R]evise and redisplay
   2 x::Foo.a
 • 2 bar(x::Foo.a::Int64)
   ↩
bar(v) in Main at REPL[3]:1
1 bar(v::Int64)::Int64 = v::Int64 + 1
Select a call to descend into or ↩ to ascend. [q]uit. [b]ookmark.
Toggles: [w]arn, [h]ide type-stable statements, [t]ype annotations, [s]yntax highlight for Source/LLVM/Native.
Show: [S]ource code, [A]ST, [T]yped code, [L]LVM IR, [N]ative code
Actions: [E]dit source code, [R]evise and redisplay
   1  v::Int64 + 1
 • ↩

julia>
```

Thanks Cthulhu: you're an amazing tool. (And thanks @vchuravy ❤️)

So I'm down to nospeicalize everything! Thoughts? I'll send up a commit like that too.

Reply:
K, i put that into a separate PR, here, since I kind of went crazy and added a bunch of them everywhere it seemed to not depend on `T`:
https://github.com/JuliaLang/julia/pull/49509

I'll merge this one now, since we're happy with it, and we can discuss what to do next on https://github.com/JuliaLang/julia/pull/49509.

Reply:
> So then: the only fear would be that if this is the **first exception** we've ever called `close()` on for this channel type, we might still stackoverflow. But if we've done it for some other exceptions first, we'll be okay.

Oh, even this isn't a concern, since when it's devirtualized, it'll be inferred during its caller's inference! Wow, julia is pretty rad. 😎